### DISCUSSION TITLE
Gapless sequential ids

### DISCUSSION THREAD

**Discussion by ItsRobbAllen** [ravendb/ravendb#15963] [Q&A] [answered]
At the most basic level, I am trying to implement event sourcing. For any stream of events, I need to ensure that each event of an aggregate root ID + version is perfectly sequential.

For example, if I have an aggregate root of `Invoices/1763-A` and am using version 1 of the aggregate root, I would like the IDs of the events to be `Invoices/1763-A/1/events/1`, `Invoices/1763-A/1/events/2`, `Invoices/1763-A/1/events/3`, etc.

If two systems try to write an event at the same time, my expectations is that both events would be written but first-come-first-numbered.

If I up the version, I'd expect the first event written would be `Invoices/1763-A/2/events/1`

I know Raven can assure sequential id's across clusters & whatnot, but they can have gaps if maybe a write failed or something. Is there anything built into Raven that would absolutely guarantee a sequential number or is this something I will need to look at handling externally?

If the latter, does anyone have any good ideas on how to ensure sequential numbers? It's critical that event 1 is followed by event 2 and there aren't gaps. If someone physically deletes an event, then that's on us and hopefully revisions will save us 😄

**Comment by georgiosd** [comment]
Hey @ItsRobbAllen - I'm curious - why is the sequential identifier important?

Don't you have to track the previous event or some sort of etag to verify correctness? If two events arrive out of order from the network from different users, presumably you need to pick one or essentially resolve it as a conflict, right?

**Comment by ayende** [comment, marked as answer]
This is a *really* complex problem.

On the face of it, fairly trivial, because RavenDB is transactional. You can have an operation that increments a value in one document and use that as the id for another.

The reason this is complex is what happens when you are running in distributed environment. Consider the case where you have two operations to add an event to the same document, on two different nodes at the same time.
What do you expect will happen then?

A better question, however, is why do you want this? What is the importance of absolute serial values?

**Reply by ItsRobbAllen** [reply]
The complexity is what I've been worrying about the entire time as we run microservices in containers, and when you have several containers working at the same time, you're eventually going to have a conflict.

However, your question is the kicker here & I think I'm conflating the "What do you need?" with "How do you do it?"

Technically speaking, it doesn't even have to be an ID. It could be a property on the document itself. The "What do you need" part is "I need to ensure events entered into the system can be ordered exactly how they were put in". Working with financial records, this is easy to visualize - if the events of a bank transaction get disordered, you could find yourself withdrawing money before you deposit it.

All that said, the only reason I can think of for the numbers having to be perfectly sequential is for human readability, which is not a good reason to add complexity to a system, especially where humans rarely will be reading.

So for now, I'm going to simply use the `|` operator & let Raven handle "Next... enough" for me & see how it goes. You can order by ID, no?

Thank you. I can't tell you how many times I've written up issues here only to delete them because by the time I finished explaining the problem in a detailed enough way for others to understand, I end up exposing the problem. The problem here is my OCD freaking out that there could be a gap in a number when to the system, it doesn't really matter.

**Reply by ayende** [reply]
Yes, that would work. For financial transactions, by the way, dates (the last modified) would also work, probably.

I'm not familiar with any financial institution that care about millisecond precision for transactions.

**Reply by georgiosd** [reply]
But regardless, is the ID enough to ensure ordering? Don't you also need to know that the user acted on the most recent set of data, i.e. having to track the previous event? Seems to me a very similar problem to optimistic concurrency in Raven.

Oh and by the way, you're not alone there @ItsRobbAllen on discovering the solution when you have to articulate the problem :)

**Reply by ItsRobbAllen** [reply]
@georgiosd 
` Don't you also need to know that the user acted on the most recent set of data, i.e. having to track the previous event?`

In Event Sourcing, the only way to know the **authoritative** current state of the system is to replay all the events in order which is done _each and every time you add an event through the aggregate root_.  Each event you create loads the entire stream*, builds the aggregate, then applies your event. If the event fails, then nothing is added to the stream

My joint bank account with my wife has $500 in it.

I go to the drive up ATM and withdraw $500.

My wife, at the same time, is in a physical bank at the teller withdrawing $500.

Both the ATM and the teller at the bank, _at the time they started counting the money_ worked on the latest set of data that said there was money available.

The ATM completes its transaction at nearly the same time as the teller. It loads my account by replaying all my transactions since the beginning of time*. The aggregate root adds everything together, sees that there is, in fact enough money for the transaction, and applies the $500 withdrawal (i.e. saves the withdrawal event to the stream)

The teller, at nearly the exact same time, presses enter on her keyboard confirming that she had the $500 needed to provide the transaction. Her screen was showing our balance as of a minute ago, but that's no longer the case.

The same thing happens though. The aggregate root is built from the entire stream, including now the ATM withdrawal transaction. The $500 withdrawal event is attempted, but the aggregate root now says "There are not enough funds to complete this!"

Now, it can either reject the transaction & that event never gets saved to the stream, OR it can record the event AND create a new `OverdraftEvent` that gets saved at the same time. <-- THIS is the scenario where problems can arrive, should another event happen between the time it creates both events, BUT... they should still be in order.

From everything I've learned about event sourcing, one takeaway is that not only is it possible for things to get mixed together, it's **probable** so best to just plan for it vs. try to tackle the impossible task of preventing it from ever happening.

*you can make snapshots along the way to prevent having to start from the very beginning, but this is irrelevant to the conversation.

**Reply by ayende** [reply]
Your scenario assumes that there is a single source that you can rely on.
Using the bank example is great. Imagine that you & your wife go to *different* branches to withdraw money.
The bank does _not_ serialize all the operations, mind. Each branch is independent. And the timestamp is _identical_.

That is how ATMs work, by the way, and the idea is that you consolidate the log at given point (end of business day) to generate the final value.

In financial transactions, for example, the _time_ of the event does *not* matter.
Here are a few good examples:
https://www.forbes.com/sites/halahtouryalai/2013/06/11/yes-banks-are-reordering-your-transactions-and-charging-overdraft-fees
https://www.nber.org/digest/202103/bank-ordering-debit-charges-and-use-payday-lenders

**Reply by ayende** [reply]
In the context of RavenDB, if you have two documents written to different nodes at the same time, representing events on the same aggregate, there is not _meaning_ to the ordering.

You can use cluster wide transactions to enforce global ordering, but should you?

Have you seen this:
https://dzone.com/articles/data-modeling-with-indexes-event-sourcing-part-1

The idea is that you can just write your events, and the RavenDB index will generate the aggregate for you, also handling ordering, etc.

**Reply by georgiosd** [reply]
Right, and in the bank example, the outcome is binary (can/cannot withdraw) - if it wasn't binary and the operation, if accumulated in the wrong order, would give a different result, that would also be a problem.

For example, multiplying after adding a value - can't think of a real scenario off the top of my head but sounds like that would get me into trouble if I didn't track what the previous state was.

**Reply by ayende** [reply]
There is a big cost for perfect consistency. In the real world, you almost never pay it.
Because business are built on high latency systems (a guy on a horse being the faster thing around for many years).
So there are a lot of compensating actions.

For example, if you buy a stock, you aren't actually buying it _right now_, it will be settled in a *couple of days*.
What actually happened is that you are submitting an order that would be accepted and processed later and the account shows the "data" as if it happened.
Makes for some interesting edge cases, see Robbinhood a year or so ago :-)

If you need perfect consistency, you pay the price for that (cluster wide transaction, single source of truth, etc).
The issue is when you want to have it and not have it at the same time.

**Reply by georgiosd** [reply]
Always fascinating to hang out in the RavenDB forums :)

**Reply by ItsRobbAllen** [reply]
Yes it is. 

In my case, there are a TON of rules & checks that have to be run and the order of each event, much like PEMDAS in math, matters to the end result. Yet, like Oren indicates, it's not possible to have _perfection_ so I'm just writing into the system ways to correct errors.

These should be few & far between as our application is not millions of operations per minute and we only have 1 cluster as it is. Still, better to be prepared.

**Reply by ayende** [reply]
FWIW, I wrote a blog post on the topic:
https://ayende.com/blog/199618-A/generating-sequential-numbers-in-a-distributed-manner?key=5c574503f5444912bf7f8a2922ecae21
